[docs] Create pickers migration guide (#23605)
mui-org/material-ui: docs/src/pages/guides/pickers-migration/pickers-migration.md

<TimePicker renderInput={(props) => <TextField {...props} />} />
```

Previously, props were spread on the `<TextField />` component. From now on you will need to use the new `renderInput` prop to provide these:

```diff
<DatePicker
- label="Date"
- helperText="Something"
+ renderInput={props => <TextField label="Date" helperText="Something" /> }
/>
```

## State management

The state/value management logic for pickers was rewritten from scratch. Pickers will now call the `onChange` prop when each view of the date picker ends is completed. The `onError` handler is also completely different. Triple-check your pickers with forms integration, because form-integration issues can be subtle.

## No required mask

Mask is no longer required. Also, if your provided mask is not valid, pickers will just ignore the mask, and allow arbitrary input.

```jsx
<DatePicker
  mask="mm"
  value={new Date()}
  onChange={console.log}
  renderInput={(props) => (
    <TextField {...props} helperText="invalid mask" />
  )}
/>

<DatePicker
  value={new Date()}
  onChange={console.log}
  renderInput={(props) => (
    <TextField {...props} helperText="valid mask" />
  )}
/>
```

## And many more

> There are a lot more changes, like prop renames, deletions, etc. Be careful to update using find & replace. Make sure your tests and build pass!

Reviewer: Lost in stale review - these need to be documented, as a list at least.

Reviewer 2: I personally think there are too many to make it worth the time of a developer to read, but let's see what Dmitriy think about it. It would close to list the changes between v0 and v1. It won't be easy.

Reviewer: > worth the time of a developer to read

How is a developer supposed to upgrade without a reference as to what has changed?

> prop renames

From what to what?

> deletions

Which props?

> etc.

Hand wavy...

Author: The project was mostly rewritten from scratch. IMHO info which props was renamed will only slow up the migration time, because it may have no sense to follow the giant step by step guide and rename/change all props.

Reviewer 2: > How is a developer supposed to upgrade without a reference as to what has changed?

He rewrite

Reviewer: So we can close this PR?

Reviewer 2: Are you saying https://material-ui.com/guides/migration-v0x/#migration-from-v0-x-to-v1 is useless?

Reviewer 2: I think that it's still better than nothing.

Reviewer: You just said that you don't expect developers to upgrade, you expect them to rewrite. If you don't expect anyone to upgrade, what's the upgrade guide for?

Reviewer: Question reamin unanswered.

Reviewer: Bump

Reviewer 2: This documentation is a step forward. Sure, it might not be perfect but it's not a reason to prevent merging it. I think that we should be grateful for the contribution that Dmitiry did. We get 80% of the value with 20% of the effort. For a first iteration, I think that it's more than enough. While upgrading might require a rewrite (hopefully not in the simplest usage of the component), the underlying concept didn't change that much. It's still interesting to have an overview of what happened.

> # Upgrading from @material-ui-pickers

Reviewer 2: To match the name of the page in the side nav?

```suggestion
# Migration from @material-ui-pickers
```

> <p class="description"> @material-ui/pickers was moved to the @material-ui/lab </p>

Reviewer: ```suggestion
<p class="description">`@material-ui/pickers` was moved to the @material-ui/lab</p>
```

> ## ☢️ Attention

Reviewer: ```suggestion
> ## ☢️ Attention
```

> Pickers were entirely changed. In most places, the logic was rewritten from scratch, so it is not possible to maintain the whole list of changes. Here is an overview of the most important concepts that were changed. If you are going to migrate – the easiest way will be to go through the component usages in your codebase and rewrite them one-by-one. Do not forget to run your tests after all!

Reviewer: ```suggestion
> Pickers were entirely changed. In most places, the logic was rewritten from scratch, so it's not possible to maintain the whole list of changes. Here's an overview of the most important concepts that were changed. If you are going to migrate, the easiest way will be to go through the component usages in your codebase and rewrite them one-by-one. Don't forget to run your tests after each!
```

> This guide is an overview of the core concepts that were changed from the pickers v3.2.10.

Reviewer: ```suggestion
This guide is an overview of the core concepts that were changed from pickers v3.2.10.
```

## ⬇️ Install

> First of all you need to install the `@material-ui/lab` if it's not installed. Nothing more required.

Reviewer: ```suggestion
You simply need to install the `@material-ui/lab` package if it's not already installed. Nothing more required.
```

> First of all you need to install the `@material-ui/lab` if it's not installed. Nothing more required.

Reviewer 2: ```suggestion
First of all, you need to install the `@material-ui/lab` if it's not installed. Nothing more is required.
```

## 🌚 Imports

> We do not expose the `keyboard` version of pickers anymore. All the version of mobile and desktop pickers implements keyboard inputs for accessibility reasons.

Reviewer: ```suggestion
The `keyboard` version of pickers is no longer published. All versions of mobile and desktop pickers implement keyboard input for accessibility.
```

```diff

> - import { KeyboardDatePicker } from '@material-ui/pickers'

Reviewer 2: ```suggestion
-import { KeyboardDatePicker } from '@material-ui/pickers';
```

> + import { DatePicker } from '@material-ui/lab'

Reviewer 2: ```suggestion
+import DatePicker from '@material-ui/lab/DatePicker';
```

> - <KeyboardDatePicker />
> + <DatePicker />

Reviewer 2: ```suggestion
-<KeyboardDatePicker />
+<DatePicker />
```

```

> Also instead of `variant` prop specific wrappers of components were moved to different imports. So you will not bundle the `Dialog` if you are using only desktop picker view.

Reviewer: ```suggestion
Also, instead of providing a `variant` prop, these were moved to different imports, meaning that your bundle won't  include `Dialog` if you are using only the desktop picker.
```

- `<DesktopDatePicker />` – Only desktop view.
- `<MobileDatePicker />` – Only mobile view.
- `<DatePicker />` – Mobile or Desktop view according to the user **pointer** preference.
- `<StaticDatePicker />` – The picker view itself, without input or any other wrapper.

```diff

> + import { DesktopDatePicker } from '@material-ui/lab'
> 
> - <DatePicker variant="inline" />
> + <DesktopDatePicker />

Reviewer 2: ```suggestion
+import DesktopDatePicker from '@material-ui/lab/DesktopDatePicker';

-<DatePicker variant="inline" />
+<DesktopDatePicker />
```

```

> The same convention for `TimePicker` – `<DesktopTimePicker>` and `<MobileTimePicker />`.

Reviewer: ```suggestion
The same convention applies to `TimePicker` – `<DesktopTimePicker>` and `<MobileTimePicker />`.
```

## MuiPickersUtilsProvider

> The `MuiPickersUtilsProvider` was removed in favor of `LocalizationProvider`. Also, pickers do not require to install date-io adapters manually. Everything comes together with the `lab`.

Reviewer: ```suggestion
The `MuiPickersUtilsProvider` was removed in favor of `LocalizationProvider`. Also, pickers do not require you to install date-io adapters manually. Everything is included with the `lab`.
```

❌ Before:

> ```jsx

Reviewer 2: ```suggestion
```js
```

import AdapterDateFns from '@date-io/date-fns';
import { MuiPickersUtilsProvider } from '@material-ui/pickers/MuiPickersUtilsProvider';
```

✅ After:

```jsx
import AdapterDateFns from '@material-ui/lab/AdapterDateFns';
import LocalizationProvider from '@material-ui/lab/LocalizationProvider';

function App() {
  return (
    <LocalizationProvider dateAdapter={AdapterDateFns}>
      ...
    </LocalizationProvider>
  )
);
```

## Render input

> We introduced a new **required** `renderInput` prop. It simplifies using different than material-ui text field input components.

Reviewer: ```suggestion
We introduced a new **required** `renderInput` prop. This simplifies using non-Material-UI text field input components.
```

> We introduced a new **required** `renderInput` prop. It simplifies using different than material-ui text field input components.

Reviewer 2: ```suggestion
We have introduced a new **required** `renderInput` prop. It simplifies using different input than Material-UI's TextField component.
```

```jsx
<DatePicker renderInput={(props) => <TextField {...props} />} />
<TimePicker renderInput={(props) => <TextField {...props} />} />
```

> Also before we were spreading props down to the `<TextField />` props. From now you need to use this new **required prop** `renderInput` to render the prop:

Reviewer: ```suggestion
Previously, props were spread on the `<TextField />` component. From now on you will need to use the new `renderInput` prop to provide these:
```

```diff

>   <DatePicker
> -   label="Date"
> -   helperText="Something"
> +   renderInput={props => <TextField label="Date" helperText="Something" /> }
>   />

Reviewer 2: ```suggestion
<DatePicker
- label="Date"
- helperText="Something"
+ renderInput={props => <TextField label="Date" helperText="Something" /> }
/>
```

```

## State management

> The state/value management logic for pickers was rewritten from scratch. They will now call the `onChange` prop when each view of picker ends its input and `onError` handler is now completely different. Triple-check your pickers with forms integration, because form-integration issues can be not obvious.

Reviewer: ```suggestion
The state/value management logic for pickers was rewritten from scratch. Pickers will now call the `onChange` prop when each view of picker ends its input. The `onError` handler is  also completely different. Triple-check your pickers with forms integration, because form-integration issues can be subtle.
```

## No required mask

> Mask is no longer required. If your passed mask is not valid for the input – pickers will just ignore the mask input and allow the arbitrary input.

Reviewer: ```suggestion
Mask is no longer required. Also, if your provided mask is not valid, pickers will just ignore the mask, and allow arbitrary input.
```

```jsx
<DatePicker
  mask="mm"
  value={new Date()}
  onChange={console.log}
  renderInput={(props) => (
    <TextField {...props} helperText="invalid mask" />
  )}
/>

<DatePicker
  value={new Date()}
  onChange={console.log}
  renderInput={(props) => (
    <TextField {...props} helperText="valid mask" />
  )}
/>
```

## And many more

> There are a lot more changes, like prop renames, deletions, etc. Be careful to update using find & replace. Make sure your tests and build passed!

Reviewer: ```suggestion
There are a lot more changes, like prop renames, deletions, etc. Be careful to update using find & replace. Make sure your tests and build passed!
```

Theses needed to be documented, at least as a list.

# Migration from @material-ui-pickers

<p class="description">@material-ui/pickers was moved to the @material-ui/lab.</p>

> > **⚠️ The date picker components were rewrote**. In most places, the logic was rewritten from scratch, so it's not possible to maintain the whole list of changes. Here's an overview of the most important concepts that were changed. If you are going to migrate, the easiest way will be to go through the component usages in your codebase and rewrite them one-by-one. Don't forget to run your tests after each!

Reviewer: ```suggestion
> **⚠️ The date picker components were rewritten**. In most places, the logic was rewritten from scratch, so it isn't possible to maintain the whole list of changes. Here's an overview of the most important concepts that were changed. If you are going to upgrade, the easiest way might be to go through each picker usage in your codebase, and rewrite them one at a time. Don't forget to run your tests after each!
```
